wolfcrypt/src/evp.c

```diff
@@ -6878,6 +6878,446 @@ int wolfSSL_EVP_get_hashinfo(const WOLFSSL_EVP_MD* evp,
+{
+    int cpysz;
+    int res;
+    word32 outsz = 0;
+
+    WOLFSSL_ENTER("wolfSSL_EVP_EncodeUpdate");
+
+    if( ctx == NULL || out == NULL || in == NULL || outl == NULL )
+        return 0;
+
+    *outl = 0;
+
+    /* if the remaining data exit in the ctx, add input data to them to create a block(48bytes) for encoding*/
+    if( ctx->remaining > 0 && inl > 0) {
+        cpysz = min( (BASE64_ENCODE_BLOCK_SIZE - ctx->remaining), inl );
+        XMEMCPY( ctx->data + ctx->remaining, in, cpysz);
+        ctx->remaining += cpysz;
+        in             += cpysz;
+        inl            -= cpysz;
+
+        /* check if a block for encoding exists in ctx.data, if so encode it */
+        if( ctx->remaining >= BASE64_ENCODE_BLOCK_SIZE ) {
+            /* Base64_Encode asks the out buff size via the 4th param*/
+            outsz = BASE64_ENCODED_BLOCK_SIZE + 1;
+            res = Base64_Encode(ctx->data, BASE64_ENCODE_BLOCK_SIZE, out, &outsz);
+            if( res == 0)
+                *outl = outsz;
+            else
+                return 0;   /* return with error */
+        }
+        else {
+            /* could not create a block */
+            *outl = 0;
+            return  1;
+        }
+    }
+    /* Here, there is no data left in ctx, so try processing the data of the specified input data.  */
+
+    while( inl >= BASE64_ENCODE_BLOCK_SIZE) {
+        outsz = BASE64_ENCODED_BLOCK_SIZE + 1;   /* 64 byte and one for LF*/
+        res = Base64_Encode( in, BASE64_ENCODE_BLOCK_SIZE,out,&outsz);
+        if( res == 0) {
+            in    += BASE64_ENCODE_BLOCK_SIZE;
+            inl   -= BASE64_ENCODE_BLOCK_SIZE;
+            out   += outsz;
+            *outl += outsz;
+        }
+        else {
+            *outl = 0;
+            return  0;
+        }
+    }
+
+    /* if remaining data exit, copy them into ctx for the next call*/
+    if( inl > 0 ) {
+        XMEMSET( ctx->data,0,sizeof(ctx->data));
+        XMEMCPY( ctx->data, in, inl);
+        ctx->remaining = inl;
+    }
+
+    return 1;   /* returns 1 on success, 0 on error */
+}
+/*  wolfSSL_EVP_EncodeFinal encodes data in ctx and outputs to out.
+ */
+void wolfSSL_EVP_EncodeFinal(WOLFSSL_EVP_ENCODE_CTX* ctx,
+                unsigned char*out, int *outl)
+{
+    word32 outsz = 0;
+    int res;
+
+    WOLFSSL_ENTER("wolfSSL_EVP_EncodeFinal");
+
+    if( outl == NULL)
+        return;
+
+    if( ctx == NULL || out == NULL  ) {
+        *outl = 0;
+        return;
+    }
+    /* process remaining data in ctx */
+    outsz = BASE64_ENCODED_BLOCK_SIZE + 1;   /* 64 byte and one for LF*/
+    res = Base64_Encode( ctx->data, ctx->remaining ,out, &outsz);
+    if( res == 0)
+        *outl = outsz;
+    else
+        *outl = 0;
+
+    ctx->remaining = 0;
+    XMEMSET( ctx->data,0,sizeof(ctx->data));
+
+    return;
+}
+#endif /* WOLFSSL_BASE64_ENCODE */
+#if defined(WOLFSSL_BASE64_DECODE)
+/* borrowed from coding.c */
+static WC_INLINE int Base64_SkipNewline(const byte* in, word32* inLen,
```

Comment:
Instead of copying this function from coding.c, you could remove the "static" declaration in coding.c, and make it a WOLFSSL_LOCAL function in <wolfssl/wolfcrypt/coding.h>, then use it here in evp.c.  Would reduce duplicate code.

Reply:
Changed to call the function in coding.c.